File: src/costs.ts
import type { Database, SqlExecutor } from "./database.js";
import { required } from "./database.js";
import { newId } from "./serialization.js";

export class BudgetExceeded extends Error {}
export class CostService {
  constructor(private readonly db: Database) {}
  async createPrice(input: Record<string, unknown>) {
    return required(
      this.db,
      `INSERT INTO price_catalog(id,provider,model,version,input_per_million,output_per_million,currency,effective_from,effective_to)
    VALUES($1,$2,$3,$4,$5,$6,$7,$8,$9) RETURNING *`,
      [
        newId(),
        input.provider,
        input.model,
        input.version,
        Number(input.input_per_million),
        Number(input.output_per_million),
        input.currency ?? "USD",
        input.effective_from,
        input.effective_to ?? null,
      ],
    );
  }
  async price(
    provider: string,
    model: string,
    executor: SqlExecutor = this.db,
  ) {
    return (
      await executor.query<Record<string, unknown>>(
        `SELECT * FROM price_catalog WHERE provider=$1 AND model=$2 AND effective_from<=now() AND (effective_to IS NULL OR effective_to>now()) ORDER BY effective_from DESC LIMIT 1`,
        [provider, model],
      )
    ).rows[0];
  }
  async estimate(
    provider: string,
    model: string,
    inputTokens: number,
    outputTokens: number,
    executor: SqlExecutor = this.db,
  ) {
    const price = await this.price(provider, model, executor);
    if (!price) return { amount: 0, currency: "USD", price: undefined };
    return {
      amount:
        (inputTokens * Number(price.input_per_million) +
          outputTokens * Number(price.output_per_million)) /
        1_000_000,
      currency: String(price.currency),
      price,
    };
  }
  async assertRunBudget(
    provider: string,
    model: string,
    inputTokens: number,
    outputTokens: number,
    limit?: number,
  ) {
    if (limit === undefined) return;
    const estimate = await this.estimate(
      provider,
      model,
      inputTokens,
      outputTokens,
    );
    if (estimate.amount > limit)
      throw new BudgetExceeded(
        `estimated run cost ${estimate.amount.toFixed(6)} ${estimate.currency} exceeds ${limit}`,
      );
  }
  async tenantSpend(tenantId: string, since: Date) {
    const row = (
      await this.db.query<{ amount: number }>(
        "SELECT COALESCE(sum(amount),0) amount FROM cost_records WHERE tenant_id=$1 AND created_at>=$2",
        [tenantId, since.toISOString()],
      )
    ).rows[0];
    return Number(row?.amount ?? 0);
  }
  async record(
    tx: SqlExecutor,
    input: {
      tenantId: string;
      runId: string;
      provider: string;
      model: string;
      requests: number;
      inputTokens: number;
      outputTokens: number;
      totalTokens: number;
      raw: unknown;
    },
  ) {
    const usage = await required<Record<string, unknown>>(
      tx,
      `INSERT INTO usage_records(id,tenant_id,run_id,model,provider,requests,input_tokens,output_tokens,total_tokens,raw_json)
    VALUES($1,$2,$3,$4,$5,$6,$7,$8,$9,$10::jsonb) RETURNING *`,
      [
        newId(),
        input.tenantId,
        input.runId,
        input.model,
        input.provider,
        input.requests,
        input.inputTokens,
        input.outputTokens,
        input.totalTokens,
        JSON.stringify(input.raw),
      ],
    );
    const estimate = await this.estimate(
      input.provider,
      input.model,
      input.inputTokens,
      input.outputTokens,
      tx,
    );
    const cost = await required<Record<string, unknown>>(
      tx,
      `INSERT INTO cost_records(id,tenant_id,run_id,usage_record_id,price_id,amount,currency,calculation_json)
    VALUES($1,$2,$3,$4,$5,$6,$7,$8::jsonb) RETURNING *`,
      [
        newId(),
        input.tenantId,
        input.runId,
        usage.id,
        estimate.price?.id ?? null,
        estimate.amount,
        estimate.currency,
        JSON.stringify({
          input_tokens: input.inputTokens,
          output_tokens: input.outputTokens,
          price_version: estimate.price?.version ?? null,
        }),
      ],
    );
    return { usage, cost };
  }
  async usage(tenantId: string, runId?: string) {
    return (
      await this.db.query(
        `SELECT * FROM usage_records WHERE tenant_id=$1 ${runId ? "AND run_id=$2" : ""} ORDER BY created_at DESC`,
        runId ? [tenantId, runId] : [tenantId],
      )
    ).rows;
  }
  async costs(tenantId: string, runId?: string) {
    return (
      await this.db.query(
        `SELECT * FROM cost_records WHERE tenant_id=$1 ${runId ? "AND run_id=$2" : ""} ORDER BY created_at DESC`,
        runId ? [tenantId, runId] : [tenantId],
      )
    ).rows;
  }
}

File: src/agent_system/costs.py
from __future__ import annotations

from datetime import UTC, datetime

from sqlalchemy import func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from .db import Database
from .models import CostRecord, PriceRecord, UsageRecord


class BudgetExceeded(RuntimeError):
    pass


class CostService:
    def __init__(self, db: Database):
        self.db = db

    async def find_price(self, provider: str, model: str) -> PriceRecord | None:
        now = datetime.now(UTC)
        async with self.db.sessions() as session:
            return await session.scalar(
                select(PriceRecord)
                .where(
                    PriceRecord.provider == provider,
                    PriceRecord.model == model,
                    PriceRecord.effective_from <= now,
                    or_(PriceRecord.effective_to.is_(None), PriceRecord.effective_to > now),
                )
                .order_by(PriceRecord.effective_from.desc())
                .limit(1)
            )

    async def estimate(
        self, provider: str, model: str, input_tokens: int, output_tokens: int
    ) -> float:
        price = await self.find_price(provider, model)
        if price is None:
            return 0.0
        return (
            input_tokens * price.input_per_million + output_tokens * price.output_per_million
        ) / 1_000_000

    async def assert_within_run_budget(
        self,
        *,
        provider: str,
        model: str,
        input_tokens: int,
        output_tokens: int,
        limit_usd: float | None,
        run_id: str | None = None,
    ) -> None:
        if limit_usd is None:
            return
        prior = await self.run_spend(run_id) if run_id else 0.0
        amount = prior + await self.estimate(provider, model, input_tokens, output_tokens)
        if amount >= limit_usd:
            raise BudgetExceeded(
                f"run cost limit reached before next model call: {amount:.6f} >= {limit_usd:.6f}"
            )

    async def run_spend(self, run_id: str) -> float:
        async with self.db.sessions() as session:
            return float(
                await session.scalar(
                    select(func.coalesce(func.sum(CostRecord.amount), 0.0)).where(
                        CostRecord.run_id == run_id
                    )
                )
                or 0.0
            )

    async def record(
        self,
        *,
        tenant_id: str,
        run_id: str,
        provider: str,
        model: str,
        requests: int,
        input_tokens: int,
        output_tokens: int,
        total_tokens: int,
        raw: dict,
    ) -> tuple[UsageRecord, CostRecord]:
        price = await self.find_price(provider, model)
        async with self.db.sessions() as session, session.begin():
            return await self.record_in_transaction(
                session,
                tenant_id=tenant_id,
                run_id=run_id,
                provider=provider,
                model=model,
                requests=requests,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                total_tokens=total_tokens,
                raw=raw,
                price=price,
            )

    async def record_in_transaction(
        self,
        session: AsyncSession,
        *,
        tenant_id: str,
        run_id: str,
        provider: str,
        model: str,
        requests: int,
        input_tokens: int,
        output_tokens: int,
        total_tokens: int,
        raw: dict,
        price: PriceRecord | None,
    ) -> tuple[UsageRecord, CostRecord]:
        amount = 0.0
        calculation = {"price_missing": price is None}
        if price is not None:
            amount = (
                input_tokens * price.input_per_million + output_tokens * price.output_per_million
            ) / 1_000_000
            calculation = {
                "price_version": price.version,
                "input_per_million": price.input_per_million,
                "output_per_million": price.output_per_million,
                "input_tokens": input_tokens,
                "output_tokens": output_tokens,
            }
        usage = UsageRecord(
            tenant_id=tenant_id,
            run_id=run_id,
            model=model,
            provider=provider,
            requests=requests,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            total_tokens=total_tokens,
            raw_json=raw,
        )
        session.add(usage)
        await session.flush()
        cost = CostRecord(
            tenant_id=tenant_id,
            run_id=run_id,
            usage_record_id=usage.id,
            price_id=price.id if price else None,
            amount=amount,
            currency=price.currency if price else "USD",
            calculation_json=calculation,
        )
        session.add(cost)
        return usage, cost

    async def tenant_spend(self, tenant_id: str, since: datetime) -> float:
        async with self.db.sessions() as session:
            return float(
                await session.scalar(
                    select(func.coalesce(func.sum(CostRecord.amount), 0.0)).where(
                        CostRecord.tenant_id == tenant_id,
                        CostRecord.created_at >= since,
                    )
                )
                or 0.0
            )
